src/sexpr-wasm.c

```diff
@@ -1,269 +1,358 @@
+STATIC_ASSERT(NUM_FLAGS == OPTIONS_LENGTH);
+
+typedef struct OptionHelp {
+  int flag;
+  const char* metavar;
+  const char* help;
+} OptionHelp;
+
+static OptionHelp s_option_help[] = {
+    {FLAG_VERBOSE, NULL, "use multiple times for more info"},
+    {FLAG_HELP, NULL, "print this help message"},
+    {FLAG_DUMP_MODULE, NULL, "print a hexdump of the module to stdout"},
+    {FLAG_OUTPUT, NULL, "<filename> file to output the generated binary format"},
+    {FLAG_SPEC, NULL,
+     "parse a file with multiple modules and assertions, like the spec tests"},
+    {FLAG_SPEC_VERBOSE, NULL, "print logging messages when running spec files"},
+    {FLAG_BR_IF, NULL, "DEPRECATED: br_if is always supported now"},
+    {FLAG_USE_LIBC_ALLOCATOR, NULL,
+     "use malloc, free, etc. instead of stack allocator"},
+    {NUM_FLAGS, NULL},
+};
+#define OPTIONS_HELP_LENGTH (ARRAY_SIZE(s_option_help) - 1)
+STATIC_ASSERT(NUM_FLAGS == OPTIONS_HELP_LENGTH);
+
+static void usage(const char* prog) {
+  printf("usage: %s [option] filename\n", prog);
+  printf("options:\n");
+  struct option* opt = &s_long_options[0];
+  int i = 0;
+  for (; opt->name; ++i, ++opt) {
+    OptionHelp* help = NULL;
+
+    int n = 0;
+    while (s_option_help[n].help) {
+      if (i == s_option_help[n].flag) {
+        help = &s_option_help[n];
+        break;
+      }
+      n++;
+    }
+
+    if (opt->val) {
+      printf("  -%c, ", opt->val);
+    } else {
+      printf("      ");
+    }
+
+    if (help && help->metavar) {
+      char buf[100];
+      snprintf(buf, 100, "%s=%s", opt->name, help->metavar);
+      printf("--%-30s", buf);
+    } else {
+      printf("--%-30s", opt->name);
+    }
+
+    if (help) {
+      printf("%s", help->help);
+    }
+
+    printf("\n");
+  }
+  exit(0);
+}
+
+static void parse_options(int argc, char** argv) {
+  int c;
+  int option_index = 0;
+#ifdef USE_MIN_PARSER
+  const char* optarg = NULL;
+  int optind = 1;
+  OptionsText optionsText[OPTIONS_LENGTH];
+  int i = 0;
+  for (; i < OPTIONS_LENGTH; i++)
  {
      struct option* opt = &s_long_options[i];
+      sprintf(optionsText[i].long_name, "--%s", opt->name);
+      if (opt->val)
+      {
+          sprintf(optionsText[i].short_name, "-%c", (char)opt->val);
+      }
+  }
+#endif
+  while (1) {
+#ifndef USE_MIN_PARSER
+    c = getopt_long(argc, argv, "vhdo:", s_long_options, &option_index);
+    if (c == -1) {
+      break;
+    }
+
+#else
+    if (optind >= argc)
+    {
+        break;
+    }
+    optarg = NULL;
+    option_index = -1;
+    const char* in_opt = argv[optind++];
+    i = 0;
+    for (; i < OPTIONS_LENGTH; i++)
    {
        const struct OptionsText* optText = &optionsText[i];
+        const struct option* opt = &s_long_options[i];
+        if (
+            strcmp(in_opt, optText->long_name) == 0 ||
+            (opt->val && strcmp(in_opt, optText->short_name) == 0)
+        )
+        {
+            option_index = i;
+            c = opt->val;

```

Comment:
Yes I was aware of this case, but since I don't know if this parser will be the final thing, I figured I wouldn't support the compressed format for the time being.